File: ProjectsInjava/ToDoDatabase/app/src/main/java/com/example/tododatabase/ToDoAdapter.java
package com.example.tododatabase;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ToDoAdapter extends RecyclerView.Adapter<ToDoAdapter.ToDoViewHolder> {

    private List<ToDo> toDoList = new ArrayList<>();

    private OnItemClickListener onItemClickListener;

    public ToDoAdapter(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void setToDoList(List<ToDo> toDoList) {
        this.toDoList = toDoList;
    }

    @NonNull
    @Override
    public ToDoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_todo, parent, false);
        return new ToDoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ToDoViewHolder holder, int position) {
        holder.bind(toDoList.get(position));
    }

    @Override
    public int getItemCount() {
        return toDoList.size();
    }

    public void addAll(List<ToDo> list) {
        toDoList.clear();
        toDoList.addAll(list);
        notifyDataSetChanged();
    }

    public void addItem(ToDo item) {
        toDoList.add(0, item);
        notifyItemInserted(0);


    }

    public void updateItem(int position, ToDo item) {
        toDoList.set(position, item);
        notifyItemChanged(position);
    }


    class ToDoViewHolder extends RecyclerView.ViewHolder {

        private TextView goalName;
        private TextView goalDesc;
        private TextView goalDate;


        ToDoViewHolder(@NonNull View itemView) {
            super(itemView);
            goalName = itemView.findViewById(R.id.goalName);
            goalDesc = itemView.findViewById(R.id.goalDesc);
            goalDate = itemView.findViewById(R.id.goalDate);

        }

        public void bind(final ToDo item) {
            goalName.setText(item.getName());
            goalDate.setText(item.getDate().toString());
            if(item.getDescription().length()  >= 10 ){
                String toShow  = item.getDescription().substring(0,10);
                goalDesc.setText(toShow + ".........");
            }
            else{
                goalDesc.setText(item.getDescription());
            }

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (onItemClickListener != null) {
                        onItemClickListener.onItemClick(getAdapterPosition(), item);
                    }
                }
            });
        }
    }

    interface OnItemClickListener {
        void onItemClick(int position, ToDo item);
    }
}

File: KotlinProjects/ToDoKotlinFile/app/src/main/java/com/example/todokotlinfile/ToDoAdapter.kt
package com.example.todokotlinfile

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import java.util.*

class ToDoAdapter(private val onItemClickListener: OnItemClickListener?) :
    RecyclerView.Adapter<ToDoAdapter.ToDoViewHolder>() {
    private var toDoList: MutableList<ToDo> = ArrayList()
    fun setToDoList(toDoList: MutableList<ToDo>) {
        this.toDoList = toDoList
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ToDoViewHolder {
        val view: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_to_do, parent, false)
        return ToDoViewHolder(view)
    }

    override fun onBindViewHolder(
        holder: ToDoViewHolder,
        position: Int
    ) {
        holder.bind(toDoList[position])
    }

    override fun getItemCount(): Int {
        return toDoList.size
    }

    fun addAll(list: List<ToDo?>?) {
        toDoList.clear()
        toDoList.addAll(list)
        notifyDataSetChanged()
    }

    fun addItem(item: ToDo) {
        toDoList.add(0, item)
        notifyItemInserted(0)
    }

    fun updateItem(position: Int, item: ToDo) {
        toDoList[position] = item
        notifyItemChanged(position)
    }

    inner class ToDoViewHolder(itemView: View) :
        ViewHolder(itemView) {
        private val goalName: TextView
        private val goalDesc: TextView
        private val goalDate: TextView
        fun bind(item: ToDo) {
            goalName.setText(item.name)
            goalDate.setText(item.date.toString())
//            if (item.description!!.length >= 10) {
//                val toShow: String = item.description.substring(0, 10)
//                goalDesc.text = "$toShow........."
//            } else {
//                goalDesc.setText(item.description)
//            }
            goalDesc.setText(item.description)
            itemView.setOnClickListener { onItemClickListener?.onItemClick(adapterPosition, item) }
        }

        init {
            goalName = itemView.findViewById(R.id.textViewName)
            goalDesc = itemView.findViewById(R.id.textViewDescription)
            goalDate = itemView.findViewById(R.id.textViewDate)
        }
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int, item: ToDo?)
    }

}